src/riak_core_bucket.erl

```diff
@@ -192,19 +197,58 @@ bucket_nval_map(Ring) ->
 default_object_nval() ->
     riak_core_bucket:n_val(riak_core_bucket_props:defaults()).
 
+-spec all_n(riak_core_ring()) -> [pos_integer(),...].
 all_n(Ring) ->
+    BucketNVals = bucket_nvals(Ring),
+    BucketTypeNVals = bucket_type_nvals(),
+    ordsets:to_list(
+        ordsets:union(BucketNVals, BucketTypeNVals)
+    ).
+
+%% @private
+-spec bucket_nvals(riak_core_ring()) -> nval_set().
+bucket_nvals(Ring) ->
     BucketNs = bucket_nval_map(Ring),
     DefaultN = default_object_nval(),
     AllN = lists:foldl(fun({_, N}, Acc) ->
                                ordsets:add_element(N, Acc)
                        end, [DefaultN], BucketNs),
     AllN.
 
+%% @private
+-spec bucket_type_nvals() -> nval_set().
+bucket_type_nvals() ->
+    riak_core_bucket_type:fold(fun bucket_type_prop_nval_fold/2, ordsets:new()).
+
+%% @private
+-spec bucket_type_prop_nval_fold({bucket_type(), bucket_props()}, nval_set()) ->
+    nval_set().
+bucket_type_prop_nval_fold({_BType, BProps}, Accum) ->
+    case lists:keyfind(active, 1, BProps) of
+        {active, true} ->
+            bucket_prop_nval_fold(BProps, Accum);
+        _ ->
+            Accum
+    end.
+
+%% @private
+-spec bucket_prop_nval_fold(bucket_props(), nval_set()) ->
+    nval_set().
+bucket_prop_nval_fold(BProps, Accum) ->
+    case lists:keyfind(n_val, 1, BProps) of
+        false ->
+            Accum;
+        {n_val, NVal} ->
+            ordsets:add_element(NVal, Accum)
+    end.
+
 name(BProps) ->
-    proplists:get_value(name, BProps).
+    {name, Name} = lists:keyfind(name, 1, BProps),
```

Comment:
Fred pointed out that these lists:keyfind replacements will fail if somehow the key (name or n_val) isn't in the list - with proplists:get_value it would just return `undefined` but not lists:keyfind will return `false` and the pattern match here won't work. There was also a concern that bucket props may actually _be_ proplists, including the strange cases of bare atoms in the list, but this appears to not be the case, as riak_core_bucket_props uses all kinds of lists:* functions that will fail if a bare atom is in the list.